distributed/nanny.py

```diff
@@ -955,34 +937,49 @@ async def run():
                         init_result_q.close()
                         await worker.finished()
                         logger.info("Worker closed")
-
-        except Exception as e:
-            logger.exception("Failed to initialize Worker")
-            init_result_q.put({"uid": uid, "exception": e})
-            init_result_q.close()
-            # If we hit an exception here we need to wait for a least one
-            # interval for the outside to pick up this message. Otherwise we
-            # arrive in a race condition where the process cleanup wipes the
-            # queue before the exception can be properly handled. See also
-            # WorkerProcess._wait_until_connected (the 2 is for good measure)
-            sync_sleep(cls._init_msg_interval * 2)
-        else:
-            try:
-                loop.run_sync(run)
-            except (TimeoutError, gen.TimeoutError):
-                # Loop was stopped before wait_until_closed() returned, ignore
+            except (tornado.util.TimeoutError, asyncio.TimeoutError):
                 pass
-            except KeyboardInterrupt:
-                # At this point the loop is not running thus we have to run
-                # do_stop() explicitly.
-                loop.run_sync(do_stop)
-            finally:
-                with suppress(ValueError):
+            except Exception as e:
+                if failure_type is not None:
+                    logger.exception(f"Failed to {failure_type} worker")
+                    init_result_q.put({"uid": uid, "exception": e})
+                    init_result_q.close()
+                    # If we hit an exception here we need to wait for a least
+                    # one interval for the outside to pick up this message.
+                    # Otherwise we arrive in a race condition where the process
+                    # cleanup wipes the queue before the exception can be
+                    # properly handled. See also
+                    # WorkerProcess._wait_until_connected (the 2 is for good
+                    # measure)
+                    sync_sleep(cls._init_msg_interval * 3)
```

Comment:
```suggestion
                if failure_type is None:
                    raise

                logger.exception(f"Failed to {failure_type} worker")
                init_result_q.put({"uid": uid, "exception": e})
                init_result_q.close()
                # If we hit an exception here we need to wait for a least
                # one interval for the outside to pick up this message.
                # Otherwise we arrive in a race condition where the process
                # cleanup wipes the queue before the exception can be
                # properly handled. See also
                # WorkerProcess._wait_until_connected (the 2 is for good
                # measure)
                sync_sleep(cls._init_msg_interval * 3)
```